routing/index_graph.hpp

```diff
@@ -29,23 +32,129 @@ class JointEdge final
 
 private:
   // Target is vertex going to for outgoing edges, vertex going from for ingoing edges.
-  Joint::Id const m_target;
-  double const m_weight;
+  Joint::Id m_target;
+  double m_weight;
+};
+
+class JointEdgeGeom final
+{
+public:
+  JointEdgeGeom() = default;
+  JointEdgeGeom(Joint::Id target, vector<RoadPoint> const & path) : m_target(target), m_path(path)
+  {
+  }
+  Joint::Id GetTarget() const { return m_target; }
+  vector<RoadPoint> const & GetPath() const { return m_path; }
+
+private:
+  // Target is vertex going to for outgoing edges, vertex going from for ingoing edges.
+  Joint::Id m_target = Joint::kInvalidId;
+  vector<RoadPoint> m_path;
+};
+
+class DirectedEdge final
+{
+public:
+  DirectedEdge() = default;
+  DirectedEdge(Joint::Id from, Joint::Id to, uint32_t featureId)
+    : m_from(from), m_to(to), m_featureId(featureId)
+  {
+  }
+
+  bool operator<(DirectedEdge const & rhs) const
+  {
+    if (m_from != rhs.m_from)
+      return m_from < rhs.m_from;
+
+    if (m_to != rhs.m_to)
+      return m_to < rhs.m_to;
+
+    return m_featureId < rhs.m_featureId;
+  }
+
+  bool operator==(DirectedEdge const & rhs) const
+  {
+    return m_from == rhs.m_from && m_to == rhs.m_to && m_featureId == rhs.m_featureId;
+  }
+
+  Joint::Id GetFrom() const { return m_from; }
+  Joint::Id GetTo() const { return m_to; }
+  uint32_t GetFeatureId() const { return m_featureId; }
+
+private:
+  Joint::Id m_from = Joint::kInvalidId;
```

Comment:
Am I right that these fields are constants actually? If so, mark them as constants. On the other hand, if you need an assignment operator, leave them as is.